Add source, test and tutorial files for the algorithm SteadyCom

shjchan opened the pull request:
[*Please include a short description of enhancement here*]
SteadyCom is a recently published algorithm for analyzing steady-state microbial communities by the Maranas group. It has been implemented completely based on COBRA Toolbox. It can be added as part of the multi-species modeling in the COBRA Toolbox.
Original paper: http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005539

Addition:
- SteadyCom functions in scr/analysis/multiSpecies
- printUptakeBoundCom.m in scr/analysis/multiSpecies as the community version of printUptakeBoundCom. It is applicable to models created using createMultiSpeciesModel.m 
- getMultiSpeciesModelId.m in scr/analysis/multiSpecies to retrieve IDs and names of the exchange reactions and metabolites in an ordered format, which are important for many purposes including SteadyCom and printUptakeBoundCom. It is also applicable to models created using createMultiSpeciesModel.m 
- test/testSteadyCom: a test file completely testing all add functions
- tutorial/SteadyCom: tutorial files for SteadyCom


**I hereby confirm that I have:**

- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [X] Selected `develop` as a target branch (top left drop-down menu)

laurentheirendt commented:
Thanks @shjchan.  May you look into why `testSteadyCom/test`failed?
```
================================================================================
Error occurred in testSteadyCom/test and it did not run to completion.

    ---------
    Error ID:
    ---------
    'MATLAB:assertion:failed'

    --------------
    Error Details:
    --------------
    Error using testSteadyCom (line 103)
    Assertion failed.
================================================================================
```

shjchan commented:
Hi @laurentheirendt , that line and other assertions after that compares the optimization results with the ref. data. It is probably because of the different solvers used.  Do you know which solver was used when the error occurred? I have updated the script to assign specific tolerances to the solvers I have tested (gurobi, ibm_cplex, glpk, mosek, matlab, dqqMinos, quadMinos). And they all past the test on my machine. (though I found a small bug in dqqMinos, which I am going to submit a pull request for that).

shjchan commented:
Hi @laurentheirendt , I think I figured out the reason for the failure. As I looked into the console output, I found that the closet previous test that change LP solve is `testSolveCobraLP`. And very luckily the last solver tested in `testSolveCobraLP` is dqqMinos. And the error in the console output looks very much alike to that I had when I tested the code with dqqMinos because `writeLPPproblem` has a small bug. See #682

laurentheirendt commented:
@shjchan, may you do a rebase here now as PR #682 is merged?

Also, some functions (e.g., `RxnList2ObjMatrix`) need proper documentation. 

May you also submit the tutorial as a separate PR? Thanks.

laurentheirendt commented on test/verifiedTests/testSteadyCom/testSteadyCom.m:
```diff
@@ -0,0 +1,186 @@
+rmdir([pwd filesep 'testSteadyComPOA'], 's')
+clear CBT_LP_SOLVER
+
+%%%%%% the section is only to avoid using ddqMinos. Should be deleted after it is fixed.
```
@shjchan, on some versions of MATLAB, using `%%` in `runtests()` causes issues. May you replace these by simple comments `%`? Thanks.

laurentheirendt commented on test/verifiedTests/testSteadyCom/testSteadyCom.m:
```diff
@@ -0,0 +1,186 @@
+if dqqMinosLP
+    changeCobraSolver('dqqMinos', 'LP');
+end
+clear SOLVERS
```
Why do you want to remove `SOLVERS` at the end of the test?

laurentheirendt commented on src/analysis/multiSpecies/SteadyCom/subroutines/setCplexParam.m:
```diff
@@ -0,0 +1,114 @@
+function LP = setCplexParam(LP, solverParams,verbFlag)
```
A similar function already exists in `src/base/solvers`. May we move this file there and eventually later combine both?

shjchan replied:
Thanks @laurentheirendt. I have removed the lines you mentioned. And I have moved `setCplexParam` and `checkSolFeas` to `src/base/solvers`. Other files in `subroutines` are very specific to the SteadyCom functions. 
In the section for `ibm_cplex` in `solveCobraLP`, assigning all parameters is actually in the to-do list (line 1205). The `setCplexParam` can have this nicely done because the Cplex parameter contains structures within structures within structures and needs quite many lines to do it.
`checkSolFeas` returns the infeasibility of solutions given the problem. Is there a function doing this in the toolbox? It is needed because solvers can sometimes return very specific solution status and there may be new status numbers in newer versions so the solution.stat may not always be the best indicator for feasibility.

shjchan commented:
Hi @laurentheirendt , to summerize:
- I have merged some functions into one function. All functions in the `SteadyCom` foler start with `SteadyCom...`. Does it look neater?
- I have moved `setCplexParam` and `checkSolFeas` to src/base/solvers.
- I have removed the lines you mentioned.
Thanks.

laurentheirendt commented:
Thanks @shjchan 🎉 

@jacekwachowiak, may you format here the documentation? Thanks.
